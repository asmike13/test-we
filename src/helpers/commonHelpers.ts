import { IBook } from "../types/book"
import { ICharacter } from "../types/character"

export const getBookId = (url: string) => url.split('books/')[1]

export const getCharacterId = (url: string) => url.split('characters/')[1]
export const getCharacterIds = (urls: string[]) => urls
	.map((url) => getCharacterId(url))
	// remove undefined ids
	.filter((url) => url)

export const getCharacterName = (character?: ICharacter) => character?.name || character?.aliases[0] || ''

export const getBookFromBookList = (bookList: IBook[], id: string) => bookList.find((book) => getBookId(book.url) === id)
export const getCharacterFromCharacterList = (characterList: ICharacter[], id?: string) => characterList.find((character) => getCharacterId(character.url) === id)
