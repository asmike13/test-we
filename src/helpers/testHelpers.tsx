/* eslint-disable prefer-const */

import { Provider } from "react-redux";
import { MemoryRouter, Routes, Route } from "react-router-dom";
import { store } from "src/store";


export const getRender = (Component: () => JSX.Element, initRoute: string[] = ['/'], path?: string) => {
	return (
		<Provider store={store}>
			<MemoryRouter initialEntries={initRoute}>
				<Routes>
					<Route path="/" element={<Component />} />
					{path && <Route path={path} element={<Component />} />}
					{!path && [...initRoute].map((r, index) => (
						<Route key={index} path={r} element={<Component />} />
					))}
				</Routes>
			</MemoryRouter>
		</Provider>
	)
}

export const waitForData = (store: any, value: any) => new Promise((res) => {
	const loop = (s: any, v: any) => {
		if (JSON.stringify(store) === JSON.stringify(value)) {
			res(true)
		} else {
			setTimeout(() => {
				loop(store, value)
			}, 300);
		}
	}

	loop(store, value)
})

export const waitForElement = (
	container: HTMLElement,
	element: string,
	removed?: boolean,
	valueToEqual?: string,
	length?: number,
) => new Promise((resolve, reject) => {
	let loopTimeout: any;
	let timeout: any;
	let elementFound: any;

	const getTextContent = (el: any) => el?.textContent || el?.value;
	let isFoundAndEqual: any;

	const loop = () => {
		elementFound = valueToEqual
			? container.querySelectorAll(element)
			: container.querySelector(element);

		const isFound = elementFound && !removed && !valueToEqual;
		const isRemoved = !elementFound && removed;
		const isFoundAndLengthEqual = isFound && length && elementFound?.length === length;

		isFoundAndEqual = valueToEqual
			&& Object.values(elementFound).find((el: any) => getTextContent(el) === valueToEqual);

		if (isRemoved || isFound || isFoundAndEqual || isFoundAndLengthEqual) {
			clearTimeout(timeout);
			resolve(isFoundAndEqual || elementFound);
		} else {
			loopTimeout = setTimeout(() => {
				loop();
			}, 100);
		}
	};

	loop();

	timeout = setTimeout(() => {
		clearTimeout(loopTimeout);
		reject(Error(`${element} ${valueToEqual ? `expect ${valueToEqual} get ${getTextContent(Object.values(elementFound)[0])}` : 'not found'}`));
	}, 3000);
});

export const waitForElementToBeRemoved = (
	container: HTMLElement,
	element: string,
) => waitForElement(
	container,
	element,
	true,
);

export const waitForElementToEqual = (
	container: HTMLElement,
	element: string,
	valueToEqual: string,
) => waitForElement(
	container,
	element,
	undefined,
	valueToEqual,
);

export const waitForElementLengthToEqual = (
	container: HTMLElement,
	element: string,
	length: number,
) => waitForElement(
	container,
	element,
	undefined,
	undefined,
	length,
);
