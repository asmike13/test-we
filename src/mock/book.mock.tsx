import { IBook } from "src/types/book";

export const bookMock = {
	authors: ['string'],
	characters: ['https://anapioficeandfire.com/api/characters/1'],
	country: 'string',
	isbn: 'string',
	mediaType: 'string',
	name: 'string',
	numberOfPages: 1,
	povCharacters: ['https://anapioficeandfire.com/api/characters/1'],
	publisher: 'string',
	released: '1996-08-01T00:00:00',
	url: 'https://anapioficeandfire.com/api/books/1',
} as IBook

export const booksMock = [
	bookMock,
	{
		...bookMock,
		url: 'https://anapioficeandfire.com/api/books/2',
	},
]