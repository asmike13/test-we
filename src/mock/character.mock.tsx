import { ICharacter } from "src/types/character";

export const characterMock = {
	id: '1',
	url: 'https://anapioficeandfire.com/api/characters/1',
	name: 'string',
	gender: 'string',
	culture: 'string',
	born: 'string',
	died: 'string',
	titles: ['string'],
	aliases: ['string'],
	father: 'string',
	mother: 'string',
	spouse: 'string',
	allegiances: ['string'],
	books: ['https://anapioficeandfire.com/api/books/1'],
	povBooks: ['https://anapioficeandfire.com/api/books/1'],
	tvSeries: ['string'],
	playedBy: ['string'],
} as ICharacter

export const charactersMock = [
	characterMock,
	{
		...characterMock,
		url: 'https://anapioficeandfire.com/api/characters/2',
	}
]