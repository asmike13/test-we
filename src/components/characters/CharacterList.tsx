import FemaleIcon from '@mui/icons-material/Female';
import MaleIcon from '@mui/icons-material/Male';
import { ListItemButton, ListItemText, ListItem, List, ListItemIcon, Paper } from "@mui/material";
import { getCharacterName } from 'src/helpers/commonHelpers';
import { ICharacter } from "src/types/character";

interface ICharacterListProps {
	status: string
	characterList: ICharacter[]
	onCharacterClick: (id?: string) => void
}

const CharacterList = ({
	status,
	characterList,
	onCharacterClick,
}: ICharacterListProps) => {
	return (
		<div className="character-list">
			<h2 className="section-title">Characters :</h2>
			<Paper elevation={3}>
				<List sx={{ maxHeight: 200, overflow: 'scroll' }} id="character-scroll-list">
					{characterList.map((character, index) => (
						<ListItem disablePadding key={index} onClick={() => onCharacterClick(character.id)}>
							<ListItemButton>
								<ListItemIcon>{character.gender === "Male" ? <MaleIcon /> : <FemaleIcon />}</ListItemIcon>
								<ListItemText primary={getCharacterName(character)} />
							</ListItemButton>
						</ListItem>
					))}
					{status === 'pending' &&
						<ListItem disablePadding>
							<ListItemButton>
								<span> Load more ...</span>
							</ListItemButton>
						</ListItem>
					}
				</List>
			</Paper>
		</div >
	)
}

export default CharacterList;
