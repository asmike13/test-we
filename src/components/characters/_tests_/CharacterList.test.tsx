import renderer from 'react-test-renderer';
import { getRender } from 'src/helpers/testHelpers';
import { charactersMock } from 'src/mock/character.mock';
import CharacterList from '../CharacterList';

const props = {
	characterList: charactersMock,
	onCharacterClick: jest.fn(),
}

it('should render as snapshot with status pending', async () => {
	const snapshot = renderer.create(getRender(() => <CharacterList status='pending' {...props} />))

	expect(snapshot).toMatchSnapshot()
})

it('should render as snapshot with status fulfilled', async () => {
	const snapshot = renderer.create(getRender(() => <CharacterList status='fulfilled' {...props} />))

	expect(snapshot).toMatchSnapshot()
})
