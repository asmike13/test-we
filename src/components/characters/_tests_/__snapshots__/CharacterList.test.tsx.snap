// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render as snapshot with status fulfilled 1`] = `
<div
  className="character-list"
>
  <h2
    className="section-title"
  >
    Characters :
  </h2>
  <div
    className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation3 css-3su884-MuiPaper-root"
  >
    <ul
      className="MuiList-root MuiList-padding css-epc8ws-MuiList-root"
      id="character-scroll-list"
    >
      <li
        className="MuiListItem-root MuiListItem-gutters css-1nxmd3h-MuiListItem-root"
        disabled={false}
        onClick={[Function]}
      >
        <div
          className="MuiListItemButton-root MuiListItemButton-gutters MuiButtonBase-root css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
          onBlur={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="button"
          tabIndex={0}
        >
          <div
            className="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="FemaleIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M17.5 9.5C17.5 6.46 15.04 4 12 4S6.5 6.46 6.5 9.5c0 2.7 1.94 4.93 4.5 5.4V17H9v2h2v2h2v-2h2v-2h-2v-2.1c2.56-.47 4.5-2.7 4.5-5.4zm-9 0C8.5 7.57 10.07 6 12 6s3.5 1.57 3.5 3.5S13.93 13 12 13s-3.5-1.57-3.5-3.5z"
              />
            </svg>
          </div>
          <div
            className="MuiListItemText-root css-tlelie-MuiListItemText-root"
          >
            <span
              className="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
            >
              string
            </span>
          </div>
          <span
            className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </div>
      </li>
      <li
        className="MuiListItem-root MuiListItem-gutters css-1nxmd3h-MuiListItem-root"
        disabled={false}
        onClick={[Function]}
      >
        <div
          className="MuiListItemButton-root MuiListItemButton-gutters MuiButtonBase-root css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
          onBlur={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="button"
          tabIndex={0}
        >
          <div
            className="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="FemaleIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M17.5 9.5C17.5 6.46 15.04 4 12 4S6.5 6.46 6.5 9.5c0 2.7 1.94 4.93 4.5 5.4V17H9v2h2v2h2v-2h2v-2h-2v-2.1c2.56-.47 4.5-2.7 4.5-5.4zm-9 0C8.5 7.57 10.07 6 12 6s3.5 1.57 3.5 3.5S13.93 13 12 13s-3.5-1.57-3.5-3.5z"
              />
            </svg>
          </div>
          <div
            className="MuiListItemText-root css-tlelie-MuiListItemText-root"
          >
            <span
              className="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
            >
              string
            </span>
          </div>
          <span
            className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </div>
      </li>
    </ul>
  </div>
</div>
`;

exports[`should render as snapshot with status pending 1`] = `
<div
  className="character-list"
>
  <h2
    className="section-title"
  >
    Characters :
  </h2>
  <div
    className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation3 css-3su884-MuiPaper-root"
  >
    <ul
      className="MuiList-root MuiList-padding css-epc8ws-MuiList-root"
      id="character-scroll-list"
    >
      <li
        className="MuiListItem-root MuiListItem-gutters css-1nxmd3h-MuiListItem-root"
        disabled={false}
        onClick={[Function]}
      >
        <div
          className="MuiListItemButton-root MuiListItemButton-gutters MuiButtonBase-root css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
          onBlur={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="button"
          tabIndex={0}
        >
          <div
            className="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="FemaleIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M17.5 9.5C17.5 6.46 15.04 4 12 4S6.5 6.46 6.5 9.5c0 2.7 1.94 4.93 4.5 5.4V17H9v2h2v2h2v-2h2v-2h-2v-2.1c2.56-.47 4.5-2.7 4.5-5.4zm-9 0C8.5 7.57 10.07 6 12 6s3.5 1.57 3.5 3.5S13.93 13 12 13s-3.5-1.57-3.5-3.5z"
              />
            </svg>
          </div>
          <div
            className="MuiListItemText-root css-tlelie-MuiListItemText-root"
          >
            <span
              className="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
            >
              string
            </span>
          </div>
        </div>
      </li>
      <li
        className="MuiListItem-root MuiListItem-gutters css-1nxmd3h-MuiListItem-root"
        disabled={false}
        onClick={[Function]}
      >
        <div
          className="MuiListItemButton-root MuiListItemButton-gutters MuiButtonBase-root css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
          onBlur={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="button"
          tabIndex={0}
        >
          <div
            className="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="FemaleIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M17.5 9.5C17.5 6.46 15.04 4 12 4S6.5 6.46 6.5 9.5c0 2.7 1.94 4.93 4.5 5.4V17H9v2h2v2h2v-2h2v-2h-2v-2.1c2.56-.47 4.5-2.7 4.5-5.4zm-9 0C8.5 7.57 10.07 6 12 6s3.5 1.57 3.5 3.5S13.93 13 12 13s-3.5-1.57-3.5-3.5z"
              />
            </svg>
          </div>
          <div
            className="MuiListItemText-root css-tlelie-MuiListItemText-root"
          >
            <span
              className="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
            >
              string
            </span>
          </div>
        </div>
      </li>
      <li
        className="MuiListItem-root MuiListItem-gutters css-1nxmd3h-MuiListItem-root"
        disabled={false}
      >
        <div
          className="MuiListItemButton-root MuiListItemButton-gutters MuiButtonBase-root css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
          onBlur={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="button"
          tabIndex={0}
        >
          <span>
             Load more ...
          </span>
        </div>
      </li>
    </ul>
  </div>
</div>
`;
